src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,178 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
 public class MyHashMap<K, V> implements MyMap<K, V> {
+    private static final int DEFAULT_CAPACITY = 16;
+    private static final short GROWTH_FACTOR = 2;
+    private int threshold = 12; //initial threshold value
+    private int size;
+    private int capacity;
+    private Node<K, V>[] table;
+
+    public MyHashMap() {
+        table = (Node<K, V>[]) new Node[DEFAULT_CAPACITY];
+        capacity = DEFAULT_CAPACITY;
+    }
 
     @Override
     public void put(K key, V value) {
+        if (size == threshold) {
+            resize();
+        }
+        putToTable(key, value, table, false);
+    }
 
+    public void putAll(Map<K, V> map) {
+        for (Map.Entry<K, V> entry : map.entrySet()) {
+            put(entry.getKey(), entry.getValue());
+        }
+    }
+
+    public boolean containsKey(K key) {
+        Node<K, V> node = table[getHash(key)];
+        while (node != null) {
+            if (Objects.equals(key, node.key)) {
+                return true;
+            }
+            node = node.next;
+        }
+        return false;
+    }
+
+    public boolean containsValue(V value) {
+        for (Node<K, V> node : table) {
+            while (node != null) {
+                if (Objects.equals(value, node.value)) {
+                    return true;
+                }
+                node = node.next;
+            }
+        }
+        return false;
     }
 
     @Override
     public V getValue(K key) {
+        Node<K, V> node = table[getHash(key)];
+        while (node != null) {
+            if (Objects.equals(node.key, key)) {
+                return node.value;
+            }
+            node = node.next;
+        }
+        return null;
+    }
+
+    /*
+    wasn't tested
+     */
+    public V remove(K key) {
+        Node<K, V> node = table[getHash(key)];
+        if (node == null) {
+            return null;
+        }
+        if (Objects.equals(key, node.key)) {
+            table[getHash(key)] = node.next;
+            return node.value;
+        }
+        Node<K, V> next = node.next;
+        while (next != null) {
+            if (Objects.equals(next.key, key)) {
+                node.next = next.next;
+                return next.value;
+            }
+            node = node.next;
+            next = next.next;
```

Comment:
you can simplify it if use `while (node != null)`

Reply:
it can't be done this way, I need this if statement and following while to remove the head of the list and element from the body of the list, respectively